linked_list/linked_list.py

```diff
@@ -5,71 +5,145 @@ class Node:
         self.value = value
         self.next = next_node
+        
 
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+      self.head = None #keep the head private.Not accessible outside class
+      self.len = 0
+      self.tail = None
+      self.max = 0
+
+      
+
+      
 
     # returns the value in the first node
     # returns None if the list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
+        return self.head.value
+
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def add_first(self, value):
-        pass
+        # next_node = self.head
+        self.head = Node(value, next_node = self.head)
+        self.len += 1 
+        self.update_max(value)
+        
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity: ?
     # Space Complexity: ?
     def search(self, value):
-        pass
+        node = self.head
+        while node != None:
+            if node.value != value:
+                node = node.next
+            else:
+                return True
+        return False
+        
+        
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        if self.head == None:
+            return 0
+        return self.len
 
+            
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
-        pass
+        node = self.head
+        counter = 0
+        if self.len < index:
+            return None
+        while node != None:
+            
+            if counter == index:
+                return node.value
+            node = node.next
+            counter +=1 
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+        print(self.len)
+        return self.get_at_index(self.len-1)
+        
+       
+
+
 
     # method that inserts a given value as a new last node in the linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def add_last(self, value):
-        pass
+        
+        if self.len == 0:
+            self.tail = Node(value, next_node = None)
+            self.head = self.tail
+            self.update_max(value)
+

```

Comment:
You only really use `tail` here, but you would potentially need to update it in `add_first` and `remove_first` etc.